src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -191,7 +192,8 @@ void ClusterTaskManager::TryScheduleInfeasibleTask() {
     bool is_infeasible;
     cluster_resource_scheduler_->GetBestSchedulableNode(
         task.GetTaskSpecification(),
-        work->PrioritizeLocalNode(),
+        /*preferred_node_id*/ work->PrioritizeLocalNode() ? self_node_id_.Binary()
+                                                          : task.GetPreferredNodeID(),
```

Comment:
How does it work if there's no preferred node id?

Reply:
Empty preferred node_id is handled specially (if empty, fallback to local node) inside `HybridSchedulingPolicy`.

Reply:
Is it possible to just pass nullopt instead? It makes sense, but I think it is pretty confusing from code to understand

Reply:
Right now, `preferred_node` is just a `std::string` in `ClusterResourceScheduler`, `RayTask`, `SchedulingOptions` and `HybridSchedulingPolicy`. Do you mean changing it to `std::optional<std::string>` globally?

Reply:
As suggested, more comments about the empty string have been added. @rkooo567